src/datanode/src/heartbeat/handler/close_region.rs

```diff
@@ -0,0 +1,243 @@
+            unreachable!("CloseRegionHandler: should be guarded by 'is_acceptable'");
+        };
+        ctx.finish();
+        let mailbox = ctx.mailbox.clone();
+        let self_ref = Arc::new(self.clone());
+
+        let RegionIdent {
+            engine,
+            table_id,
+            catalog,
+            schema,
+            table,
+            region_number,
+        } = region_ident;
+
+        common_runtime::spawn_bg(async move {
+            let result = self_ref
+                .close_region_inner(
+                    engine,
+                    table_id,
+                    &TableReference::full(&catalog, &schema, &table),
+                    vec![region_number],
+                )
+                .await;
+
+            if let Err(e) = mailbox
+                .send((meta, CloseRegionHandler::map_result(result)))
+                .await
+            {
+                error!(e;"Failed to send reply to mailbox");
+            }
+        });
+
+        Ok(())
+    }
+}
+
+impl CloseRegionHandler {
+    pub fn new(
+        catalog_manager: CatalogManagerRef,
+        table_engine_manager: TableEngineManagerRef,
+    ) -> Self {
+        Self {
+            catalog_manager,
+            table_engine_manager,
+        }
+    }
+
+    fn map_result(result: Result<bool>) -> InstructionReply {
+        result.map_or_else(
+            |error| {
+                InstructionReply::CloseRegion(SimpleReply {
+                    result: false,
+                    error: Some(error.to_string()),
+                })
+            },
+            |result| {
+                InstructionReply::CloseRegion(SimpleReply {
+                    result,
+                    error: None,
+                })
+            },
+        )
+    }
+
+    /// Returns true if table has been closed.
+    async fn check_table(
+        &self,
+        catalog_name: &str,
+        schema_name: &str,
+        table_name: &str,
+        region_numbers: &[RegionNumber],
+    ) -> Result<bool> {
+        if let Some(table) = self
+            .catalog_manager
+            .table(catalog_name, schema_name, table_name)
+            .await
+            .context(error::AccessCatalogSnafu)?
+        {
+            for r in region_numbers {
+                let region_exist =
+                    table
+                        .contains_region(*r)
+                        .await
+                        .with_context(|_| error::CheckRegionSnafu {
+                            table_name: format_full_table_name(
+                                catalog_name,
+                                schema_name,
+                                table_name,
+                            ),
+                            region_number: *r,
+                        })?;
+                if region_exist {
+                    return Ok(false);
+                }
+            }
+        }
+        // Returns true if table not exist
```

Comment:
```suggestion
        // Returns true if all the regions are not present in this table.
```

and replace the docs of this function to this comment